### PR TITLE
rebar3: 3.6.1 -> 3.9.0

### PR DESCRIPTION
**Pull request by k32**
###### Motivation for this change
- Update package
- Simplify build

###### Things done

 - Updated rebar3 package
 - Removed incompatible hermetic patch (nix-build runs with network isolation, hence the hermetic patch is not needed anymore)
 - Started eliminating the need of hex pac

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
cc @sjmackenzie for review

**Comment by k32** [comment]
Anyone?  @gleber @couchemar @sjmackenzie @the-kenny

**Comment by Mic92** [comment]
```
@GrahamcOfBorg build hex2nix relxExe
```

**Comment by Mic92** [comment]
Are `hex2nix` and `relxExe` known to fail?

**Comment by k32** [comment]
Could be something related to the hermetic patch. I'll take a look later.

**Comment by k32** [comment]
Ok, these errors were caused by not updating hex cache, which is completely useless when building with nix-build. My workaround with unpacking contents of hex packages to `_checkouts` directory prior to running rebar3 seems to work just fine, so I think I'll proceed with changing the rest of the packages to use the same trick. ETA: weekend

**Comment by k32** [comment]
Unfortunately hashes used by hex are different from what could be used by nix. They concatenate the source tarball together with some metadata, which is unknown in advance [1]. It is still possible to use hex-style hash in a fixed-output derivation, but via a hack so dirty I don't even want to talk about. So I had to resort to a fixed-output derivations containing downloaded sources. It is used by Rust people, however this practice is disputed [2]

[1] https://github.com/hexpm/hex_core/blob/bfd012f7200091cac2901850b2f7a828f7bf6301/src/hex_tarball.erl#L244
[2] https://github.com/NixOS/nix/issues/2270

**Comment by k32** [comment]
@GrahamcOfBorg build hex2nix relxExe

**Comment by k32** [comment]
@Mic92 Sorry for the delay. Both hex2nix and relxExe build now and nox doesn't report any errors.

File: pkgs/development/beam-modules/default.nix
Comment (ankhers) [inline pkgs/development/beam-modules/default.nix]: These need the `hermeticRebar3` option. It defaults to true, which means `rebar3-open` and `rebar3` are now the same things. And users of `rebar3-open` will be unable to download packages from hex because it is hermetic.

File: pkgs/development/beam-modules/fetch-rebar-deps.nix
Comment (ankhers) [inline pkgs/development/beam-modules/fetch-rebar-deps.nix]: For what it's worth, this will only work on `rebar3-open` because it is not hermetic. How does this really differ from the current `fetchHex` (aside from it possibly getting non-hex deps)?

File: pkgs/development/beam-modules/default.nix
Comment (ankhers) [reply pkgs/development/beam-modules/default.nix]: After going through it a little further, I see that you removed the hermetic patch, which means (I think) it now goes against how Nix is expecting things. It is free to download anything that it wants instead of you explicitly saying what the applications dependencies are. I would like someone with more knowledge than myself to say whether this is acceptable or not.

Comment (k32) [reply pkgs/development/beam-modules/default.nix]: Hi,

In short, removal of the hermetic patch isn't against Nix's way of doing things: nix-build runs with network isolation (unless it's a fixed-output derivation) therefore rebar3 won't be able to download anything during the process of build.

Fetching dependencies is a different story. It was moved to a separate step. Using `rebar3 get-deps` produces fixed output (rebar3 itself uses sha256 to freeze the dependencies), so this is no worse than using `fetchurl`.

I described my reasoning more in depth here: https://github.com/NixOS/nixpkgs/issues/53834

File: pkgs/development/beam-modules/fetch-rebar-deps.nix
Comment (k32) [reply pkgs/development/beam-modules/fetch-rebar-deps.nix]: The current way of dealing erlang dependencies is not viable. Not only hermetic patch breaks rebar3 for everyone, but it also heavily relies on undocumented rebar3 features and file locations.

With the current approach in order to update a package one needs to

1) Ask @gleber to update hex index cache repo (note that hex index is not actually needed. it's only used to avert rebar3 errors)
2) Wait until index hash repo is merged and the derivation is updated in nixpkgs repo
3) Ping someone to run hex2nix and update hex packages in this repo. 
4) Wait again until the new PR is reviewed and merged
5) Manually convert lock file contents to the Nix expression
6) Pray that nothing doesn't have non-trivial build steps
8) Pray especially hard that no one uses rebar3 `override` feature

With the new approach every package can be maintained basically independently.

**Comment by ankhers** [comment]
Since this is still open, would you mind bumping ot 3.9.0?

**Comment by k32** [comment]
@ankhers Done, also squashed commits to reduce noise

**Comment by ankhers** [comment]
I'm good with this. My only other suggestion would be to drop `rebar3-open` since it is effectively useless. But we can do that in another PR if preferred.

**Comment by k32** [comment]
Someone may use it. Is there a good way to deprecate a package?

**Comment by ankhers** [comment]
Based on a conversation in IRC, there doesn't seem to be a way to deprecate things. We may just need to change some docs the refer to it. Lets keep it for now and we can clear it out in a different PR.

**Comment by k32** [comment]
If possible, I would like to update docs through a separate PR after the rest of planned changes are proven viable and implemented. After that both old `hex2nix` packages and `rebar3-open` can be nuked.

**Comment by k32** [comment]
@ankhers May I have this one approved so I can move on to the next step?

**Review by ankhers — APPROVED** [review]
This should be good.

**Comment by ankhers** [comment]
I can't actually merge. Will need to wait for someone else to do that.

**Pull request merged**
